service/s3/s3crypto/envelope.go

```diff
@@ -0,0 +1,137 @@
+package s3crypto
+
+import (
+	"bytes"
+	"encoding/json"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/client"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+// Envelope encryption starts off by generating a random symmetric key using AES CBC,
+// AES CTR, or AES GCM. We then generate a random IV based off the encryption cipher
+// chosen. The master key that was provided, whether by the user or KMS, will be used
+// to encrypt the randomly generated symmetric key and base64 encode the iv. This will
+// allow for decryption of that same data later.
+type Envelope struct {
+	// IV is the randomly generated IV base64 encoded.
+	IV string `json:"x-amz-iv"`
+	// CipherKey is the randomly generated cipher key.
+	CipherKey string `json:"x-amz-key-v2, x-amz-key"`
+	// MaterialDesc is a description to distinguish from other envelopes.
+	MatDesc               string `json:"x-amz-matdesc"`
+	WrapAlg               string `json:"x-amz-wrap-alg"`
+	CEKAlg                string `json:"x-amz-cek-alg"`
+	TagLen                string `json:"x-amz-tag-len"`
+	UnencryptedMD5        string `json:"x-amz-unencrypted-content-md5"`
+	UnencryptedContentLen string `json:"x-amz-unencrypted-content-length"`
+	version               int
+}
+
+func (client *Client) getEnvelope(input *s3.GetObjectInput, r *request.Request) (*Envelope, error) {
+	if value := r.HTTPResponse.Header.Get("x-amz-meta-x-amz-key-v2"); value != "" {
+		return getV2Envelope(r)
+	} else if value = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-key"); value != "" {
+		return getV1Envelope(r)
+	} else {
+		return getFromInstructionFile(input, r)
+	}
+}
+
+// TODO:
+// Need MD5 and SHA256 io.Reader
+// Need to redo all io.Reader interface for encrypt and decrypt
+// Implement KMS and Private Key RSA
+func getV2Envelope(r *request.Request) (*Envelope, error) {
+	env := &Envelope{}
+	env.CipherKey = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-key-v2")
+	env.IV = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-iv")
+	env.MatDesc = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-matdesc")
+	env.WrapAlg = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-wrap-alg")
+	env.CEKAlg = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-cek-alg")
+	env.TagLen = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-tag-len")
+	env.UnencryptedMD5 = r.HTTPResponse.Header.Get("x-amz-unencrypted-content-md5")
+	env.UnencryptedContentLen = r.HTTPResponse.Header.Get("x-amz-unencrypted-content-length")
+	env.version = 2
+	return env, nil
+}
+
+func getV1Envelope(r *request.Request) (*Envelope, error) {
+	env := &Envelope{}
+	env.CipherKey = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-key")
+	env.IV = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-iv")
+	env.MatDesc = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-matdesc")
+	env.version = 1
+	return env, nil
+}
+
+func getFromInstructionFile(input *s3.GetObjectInput, r *request.Request) (*Envelope, error) {
+	return &Envelope{}, nil
```

Comment:
Naw, just grabs the V1 envelope from the headers. However, we no longer support v1 reads due to padding attacks. So, may be best to remove the method